zcash_client_sqlite/src/wallet/init/migrations/fix_bad_change_flagging.rs

```diff
@@ -0,0 +1,198 @@
+            },
+            fees::{standard, DustOutputPolicy},
+            wallet::WalletTransparentOutput,
+        },
+        zcash_primitives::{
+            block::BlockHash,
+            transaction::{
+                components::{OutPoint, TxOut},
+                fees::StandardFeeRule,
+            },
+        },
+        zcash_protocol::value::Zatoshis,
+    };
+
+    #[test]
+    fn migrate() {
+        test_migrate(&[super::MIGRATION_ID]);
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    fn shield_transparent<T: ShieldedPoolTester>() {
+        let ds_factory = TestDbFactory::new(super::DEPENDENCIES.to_vec());
+        let cache = BlockCache::new();
+        let mut st = TestBuilder::new()
+            .with_data_store_factory(ds_factory)
+            .with_block_cache(cache)
+            .with_account_from_sapling_activation(BlockHash([0; 32]))
+            .build();
+
+        let account = st.test_account().cloned().unwrap();
+        let dfvk = T::test_account_fvk(&st);
+
+        let uaddr = st
+            .wallet()
+            .get_current_address(account.id())
+            .unwrap()
+            .unwrap();
+        let taddr = uaddr.transparent().unwrap();
+
+        // Ensure that the wallet has at least one block
+        let (h, _, _) = st.generate_next_block(
+            &dfvk,
+            AddressType::Internal,
+            Zatoshis::const_from_u64(50000),
+        );
+        st.scan_cached_blocks(h, 1);
+
+        let utxo = WalletTransparentOutput::from_parts(
+            OutPoint::fake(),
+            TxOut {
+                value: Zatoshis::const_from_u64(100000),
+                script_pubkey: taddr.script(),
+            },
+            Some(h),
+        )
+        .unwrap();
+
+        let res0 = st.wallet_mut().put_received_transparent_utxo(&utxo);
+        assert_matches!(res0, Ok(_));
+
+        let fee_rule = StandardFeeRule::Zip317;
+
+        let input_selector = GreedyInputSelector::new(
+            standard::SingleOutputChangeStrategy::new(fee_rule, None, T::SHIELDED_PROTOCOL),
+            DustOutputPolicy::default(),
+        );
+
+        let txids = st
+            .shield_transparent_funds(
+                &input_selector,
+                Zatoshis::from_u64(10000).unwrap(),
+                account.usk(),
+                &[*taddr],
+                1,
+            )
+            .unwrap();
+        assert_eq!(txids.len(), 1);
+
+        let tx = st.get_tx_from_history(*txids.first()).unwrap().unwrap();
+        assert_eq!(tx.spent_note_count(), 1);
+        assert!(tx.has_change());
+        assert_eq!(tx.received_note_count(), 0);
+        assert_eq!(tx.sent_note_count(), 0);
+        assert!(tx.is_shielding());
+
+        // Complete the migration
+        init_wallet_db(st.wallet_mut().db_mut(), None).unwrap();
+
+        // Ensure that the transaction metadata is still correct after the update produced by scanning.
```

Comment:
What scanning update? AFAICT this test doesn't do one; it just checks that the unscanned shielding transaction's metadata is not altered by the migration. It might indeed be useful to have two tests for this migration: one that ensures correct data isn't altered, and one that ensures incorrect data gets fixed.

Reply:
Copy/paste error - this comment was correct in the test for the change from `IS_NULL` to `MAX`, and the test is the same.

Reply:
[force-pushed](https://github.com/zcash/librustzcash/compare/0d82e1e..d75309080106639920f8302e555f482dfc35e6c2) to address comment.